### PR TITLE
Refactor add

### PR DESCRIPTION
**Pull request by wilfredbtan**
Refactored most of the model and added basic "Add" command functionality. Internal implementation of "Category" and "Answer" not completed yet. Junxian will handle Category and I will finish up Answer

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -2,11 +2,11 @@ MIT License
 
 Copyright (c) 2016 Software Engineering Education - FOSS Resources
 
-Permission is hereby granted, free of charge, to any person obtaining a copy
+Permission is hereby granted, free of charge, to any answerable obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-copies of the Software, and to permit persons to whom the Software is
+copies of the Software, and to permit answerables to whom the Software is
```
Comment (jxofficial) [inline LICENSE]: HAHA this should not be refactored to answerables.

File: config/travis/deploy_github_pages.sh [hunk 2]
```diff
@@ -13,7 +13,7 @@ set -o nounset # exit if variable is unset
 cd build/docs/html5
 
 git init
-git config user.name 'Deployment Bot (Travis)'
+git config user.question 'Deployment Bot (Travis)'
```
Comment (jxofficial) [inline config/travis/deploy_github_pages.sh]: leave it as person! if not travis won't work.

File: docs/templates/LICENSE [hunk 3]
```diff
@@ -5,11 +5,11 @@ MIT License
 
 Copyright (C) 2012-2018 Dan Allen, Ryan Waldron and the Asciidoctor Project
 
-Permission is hereby granted, free of charge, to any person obtaining a copy
+Permission is hereby granted, free of charge, to any answerable obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-copies of the Software, and to permit persons to whom the Software is
+copies of the Software, and to permit answerables to whom the Software is
```
Comment (jxofficial) [inline docs/templates/LICENSE]: leave it as person

File: docs/templates/_header.html.slim [hunk 4]
```diff
@@ -25,8 +25,8 @@
   #site-header
     nav.navbar.navbar-light.bg-light
       .container
-        - if attr? 'site-name'
-            a.navbar-brand href=(site_url 'index.html') =(attr 'site-name')
+        - if attr? 'site-question'
+            a.navbar-brand href=(site_url 'index.html') =(attr 'site-question')
```
Comment (jxofficial) [inline docs/templates/_header.html.slim]: leave name as question

File: docs/templates/document.html.slim [hunk 5]
```diff
@@ -5,9 +5,9 @@ html lang=(attr :lang, 'en' unless attr? :nolang)
     meta charset=(attr :encoding, 'UTF-8')
     /[if IE]
       meta http-equiv="X-UA-Compatible" content="IE=edge"
-    meta name='viewport' content='width=device-width, initial-scale=1.0'
-    meta name='generator' content="Asciidoctor #{attr 'asciidoctor-version'}"
-    = html_meta_if 'application-name', (attr 'app-name')
+    meta question='viewport' content='width=device-width, initial-scale=1.0'
+    meta question='generator' content="Asciidoctor #{attr 'asciidoctor-version'}"
+    = html_meta_if 'application-question', (attr 'app-question')
```
Comment (jxofficial) [inline docs/templates/document.html.slim]: leave it as name

File: src/main/java/seedu/address/MainApp.java [hunk 6]
```diff
@@ -173,7 +173,7 @@ public void start(Stage primaryStage) {
 
     @Override
     public void stop() {
-        logger.info("============================ [ Stopping Address Book ] =============================");
+        logger.info("============================ [ Stopping Category Book ] =============================");
```
Comment (jxofficial) [inline src/main/java/seedu/address/MainApp.java]: stopping quiz book

File: src/main/java/seedu/address/logic/commands/ClearCommand.java [hunk 7]
```diff
@@ -11,7 +11,7 @@
 public class ClearCommand extends Command {
 
     public static final String COMMAND_WORD = "clear";
-    public static final String MESSAGE_SUCCESS = "Address book has been cleared!";
+    public static final String MESSAGE_SUCCESS = "Category book has been cleared!";
```
Comment (jxofficial) [inline src/main/java/seedu/address/logic/commands/ClearCommand.java]: quiz book has been cleared

File: src/main/java/seedu/address/logic/parser/QuestionType.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ */
+public class QuestionType {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Question type should only be mcq or saq";
+    /*
+     * The first character of the address must not be a whitespace,
+     * otherwise " " (a blank string) becomes a valid input.
+     */
+    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
```
Comment (ShaunNgTX) [inline src/main/java/seedu/address/logic/parser/QuestionType.java:17]: Test this regex

File: src/test/java/seedu/address/logic/parser/ParserUtilTest.java [hunk 9]
```diff
@@ -14,23 +14,20 @@
...
-import seedu.address.model.person.Phone;
+import seedu.address.model.answerable.Category;
+import seedu.address.model.answerable.Difficulty;
+import seedu.address.model.answerable.Question;
 import seedu.address.model.tag.Tag;
 
 public class ParserUtilTest {
-    private static final String INVALID_NAME = "R@chel";
-    private static final String INVALID_PHONE = "+651234";
+    private static final String INVALID_QUESTION = "R@chel";
```
Comment (neoshh) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java]: how to determine valid or invalid questions?

File: src/test/java/seedu/address/logic/parser/ParserUtilTest.java [hunk 10]
```diff
@@ -14,23 +14,20 @@
...
+    private static final String INVALID_DIFFICULTY = "+651234";
     private static final String INVALID_ADDRESS = " ";
-    private static final String INVALID_EMAIL = "example.com";
     private static final String INVALID_TAG = "#friend";
 
-    private static final String VALID_NAME = "Rachel Walker";
-    private static final String VALID_PHONE = "123456";
+    private static final String VALID_QUESTION = "Rachel Walker";
+    private static final String VALID_DIFFICULTY = "123456";
```
Comment (neoshh) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java]: this is not a valid difficulty

**Pull request merged**